Properties

ljeub-pometry opened the pull request:
### What changes were proposed in this pull request?

Unify the properties api and create Properties and StaticProperties objects

### Why are the changes needed?

Make working with properties easier to use and properties api easier to implement

### Does this PR introduce any user-facing change? If yes is this documented?

yes, all the properties apis are different. Rudimentary docstrings are there but need to be improved

### How was this patch tested?

### Issues

closes #957, #956 

### Are there any further changes required?

miratepuffin reviewed (changes requested):
A couple of small points, mainly on documentation:

- Seems we have lost a lot of python docs
- We need to manually add the properties object to the docs rust

ljeub-pometry commented on raphtory/src/algorithms/triangle_count.rs:
```diff
@@ -67,6 +67,7 @@ pub fn triangle_count<G: GraphViewOps>(g: &G, threads: Option<usize>) -> usize {
     let step1 = ATask::new(move |s| {
         for t in s.neighbours() {
             if s.id() > t.id() {
+                println!("adding {} to neighbours of {}", s.id(), t.id());
```
Yes, this was from testing the subgraph bug, should not be here obviously...

ljeub-pometry commented on raphtory/src/db/api/properties/props.rs:
```diff
@@ -0,0 +1,80 @@
+        self.get(key).is_some()
+    }
+
+    pub fn keys(&self) -> impl Iterator<Item = LockedView<String>> + '_ {
```
yes

ljeub-pometry commented on raphtory/src/db/graph/vertices.rs:
```diff
@@ -72,116 +74,12 @@ impl<G: GraphViewOps> VertexViewOps for Vertices<G> {
-    /// An iterator of the static properties of the vertices
-    fn static_properties(&self) -> Self::ValueType<HashMap<String, Prop>> {
-        self.iter().static_properties()
+    fn properties(&self) -> Self::ValueType<Properties<VertexView<G>>> {
```
should indeed have it as it is different from the default docstring of the trait

ljeub-pometry commented on raphtory/src/db/api/view/vertex.rs:
```diff
@@ -223,56 +142,10 @@ pub trait VertexListOps:
-    /// An iterator of the values of the given property name including the times when it changed
-    /// as a vector of tuples of the form (time, property).
-    fn property_history(self, name: String) -> Self::IterType<Vec<(i64, Prop)>>;
+    /// Returns an iterator over properties of the vertices
```
Agreed, we should do that when we refactor the internal implementations

ljeub-pometry commented:
I cleaned up the minor comments and added python docs for the base types. Still need to figure out the best way to document the lists and lists of lists variants
